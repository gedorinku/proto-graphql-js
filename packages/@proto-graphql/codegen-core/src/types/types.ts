import { ProtoEnum, ProtoField, ProtoMessage, ProtoRegistry } from "@proto-graphql/proto-descriptors";
import assert from "assert";
import { FieldDescriptorProto } from "google-protobuf/google/protobuf/descriptor_pb";
import { DslFile } from "./DslFile";
import { EnumType } from "./EnumType";
import { InputObjectType } from "./InputObjectType";
import { InterfaceType } from "./InterfaceType";
import { ObjectType } from "./ObjectType";
import { OneofUnionType } from "./OneofUnionType";
import { ScalarType } from "./ScalarType";
import { SquashedOneofUnionType } from "./SquashedOneofUnionType";
import {
  exceptRequestOrResponse,
  GenerationParams,
  isIgnoredField,
  isIgnoredType,
  isInterface,
  isScalar,
  isSquashedUnion,
} from "./util";

export type GlType =
  | ScalarType
  | ObjectType
  | InterfaceType
  | OneofUnionType
  | SquashedOneofUnionType
  | EnumType
  | InputObjectType;

export function collectTypesFromFile(file: DslFile, registry: ProtoRegistry) {
  const [msgs, enums] = file.proto.collectTypesRecursively();

  return [
    ...buildObjectTypes(msgs, file, registry),
    ...buildInputObjectTypes(msgs, file, registry),
    ...buildInterfaceType(msgs, file, registry),
    ...buildSquashedOneofUnionTypes(msgs, file),
    ...buildOneofUnionTypes(msgs, file),
    ...buildEnumTypes(enums, file),
  ];
}

function buildObjectTypes(msgs: ProtoMessage[], file: DslFile, registry: ProtoRegistry): ObjectType[] {
  return msgs
    .filter((m) => !isIgnoredType(m))
    .filter((m) => !isSquashedUnion(m))
    .filter((m) => !isInterface(m))
    .filter(exceptRequestOrResponse(registry))
    .map((m) => new ObjectType(m, file));
}

function buildInputObjectTypes(msgs: ProtoMessage[], file: DslFile, registry: ProtoRegistry): InputObjectType[] {
  return msgs
    .filter((m) => !isIgnoredType(m))
    .filter(exceptRequestOrResponse(registry))
    .map((m) => new InputObjectType(m, file))
    .flatMap((t) => (file.options.partialInputs && t.hasPartialInput() ? [t, t.toPartialInput()] : t));
}

function buildInterfaceType(msgs: ProtoMessage[], file: DslFile, registry: ProtoRegistry): InterfaceType[] {
  return msgs
    .filter((m) => !isIgnoredType(m))
    .filter((m) => !isSquashedUnion(m))
    .filter((m) => isInterface(m))
    .filter(exceptRequestOrResponse(registry))
    .map((m) => new InterfaceType(m, file));
}

function buildSquashedOneofUnionTypes(msgs: ProtoMessage[], file: DslFile): SquashedOneofUnionType[] {
  return msgs
    .filter((m) => !isIgnoredType(m))
    .filter((m) => isSquashedUnion(m))
    .map((m) => new SquashedOneofUnionType(m, file));
}

function buildOneofUnionTypes(msgs: ProtoMessage[], file: DslFile): OneofUnionType[] {
  return msgs
    .filter((m) => !isSquashedUnion(m))
    .flatMap((m) => m.oneofs)
    .filter((o) => !isIgnoredField(o))
    .map((o) => new OneofUnionType(o, file));
}

function buildEnumTypes(enums: ProtoEnum[], file: DslFile): EnumType[] {
  return enums.filter((e) => !isIgnoredType(e)).map((e) => new EnumType(e, file));
}

export function getObjectFieldType(
  proto: ProtoField,
  opts: GenerationParams & { dsl: "nexus" | "pothos" }
): ScalarType | EnumType | ObjectType | InterfaceType | SquashedOneofUnionType {
  return detectType<ObjectType | InterfaceType | SquashedOneofUnionType | ScalarType>(proto, opts, (msg, file) => {
    if (isInterface(msg)) return new InterfaceType(msg, file);
    if (isSquashedUnion(msg)) return new SquashedOneofUnionType(msg, file);
    return new ObjectType(msg, file);
  });
}

export function getInputObjectFieldType(
  proto: ProtoField,
  opts: GenerationParams & { dsl: "nexus" | "pothos" }
): ScalarType | EnumType | InputObjectType {
  return detectType<InputObjectType>(proto, opts, (msg, file) => {
    return new InputObjectType(msg, file);
  });
}

function detectType<T extends ObjectType | InterfaceType | SquashedOneofUnionType | InputObjectType | ScalarType>(
  proto: ProtoField,
  opts: GenerationParams & { dsl: "nexus" | "pothos" },
  f: (msg: ProtoMessage, file: DslFile) => T
): ScalarType | EnumType | T {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const pbtype = proto.descriptor.getType()!;
  switch (pbtype) {
    case FieldDescriptorProto.Type.TYPE_STRING:
      return new ScalarType(proto, "String", opts);
    case FieldDescriptorProto.Type.TYPE_DOUBLE:
    case FieldDescriptorProto.Type.TYPE_FLOAT:
      return new ScalarType(proto, "Float", opts);
    case FieldDescriptorProto.Type.TYPE_INT64:
      return new ScalarType(proto, opts.longNumber, opts);
    case FieldDescriptorProto.Type.TYPE_UINT64:
      return new ScalarType(proto, opts.longNumber, opts);
    case FieldDescriptorProto.Type.TYPE_INT32:
      return new ScalarType(proto, "Int", opts);
    case FieldDescriptorProto.Type.TYPE_FIXED64:
      return new ScalarType(proto, opts.longNumber, opts);
    case FieldDescriptorProto.Type.TYPE_FIXED32:
      return new ScalarType(proto, "Int", opts);
    case FieldDescriptorProto.Type.TYPE_UINT32:
      return new ScalarType(proto, "Int", opts);
    case FieldDescriptorProto.Type.TYPE_SFIXED32:
      return new ScalarType(proto, "Int", opts);
    case FieldDescriptorProto.Type.TYPE_SFIXED64:
      return new ScalarType(proto, opts.longNumber, opts);
    case FieldDescriptorProto.Type.TYPE_SINT32:
      return new ScalarType(proto, "Int", opts);
    case FieldDescriptorProto.Type.TYPE_SINT64:
      return new ScalarType(proto, opts.longNumber, opts);
    case FieldDescriptorProto.Type.TYPE_BOOL:
      return new ScalarType(proto, "Boolean", opts);
    case FieldDescriptorProto.Type.TYPE_GROUP:
      throw "not supported";
    case FieldDescriptorProto.Type.TYPE_BYTES:
      throw "not supported";
    case FieldDescriptorProto.Type.TYPE_ENUM:
      assert(proto.type && proto.type.kind === "Enum");
      return new EnumType(proto.type, new DslFile(proto.type.file, opts));
    case FieldDescriptorProto.Type.TYPE_MESSAGE: {
      assert(proto.type && proto.type.kind === "Message");
      const msg = proto.type;
      if (isScalar(msg, opts)) {
        return new ScalarType(proto, opts.typeMappings[msg.fullName.toString()] as any, opts);
      }
      const file = new DslFile(msg.file, opts);
      return f(msg, file);
    }
    /* istanbul ignore next */
    default: // eslint-disable-line
      const _exhaustiveCheck: never = pbtype;
      throw "unreachable";
  }
}
