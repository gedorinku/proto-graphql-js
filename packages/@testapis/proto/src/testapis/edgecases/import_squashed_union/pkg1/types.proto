syntax = "proto3";

package testapis.edgecases.import_squashed_union.pkg1;

import "graphql/schema.proto";

message Message {
  string body = 1;
}

message SquashedOneof {
  option (graphql.object_type).squash_union = true;

  oneof message {
    Message msg = 1;
  }
}
