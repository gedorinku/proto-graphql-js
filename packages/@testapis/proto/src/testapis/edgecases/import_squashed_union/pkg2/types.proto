syntax = "proto3";

package testapis.edgecases.import_squashed_union.pkg2;

import "testapis/edgecases/import_squashed_union/pkg1/types.proto";

message Message {
  testapis.edgecases.import_squashed_union.pkg1.SquashedOneof msg = 1;
}
