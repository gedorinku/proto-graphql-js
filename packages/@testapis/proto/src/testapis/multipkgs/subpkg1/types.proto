syntax = "proto3";

package testapis.multipkgs.subpkg1;

message SubpkgMessage {
  string body = 1;
}

enum SubpkgEnum {
  SUBPKG_ENUM_UNSPECIFIED = 0;
  FOO = 1;
  BAR = 2;
}
