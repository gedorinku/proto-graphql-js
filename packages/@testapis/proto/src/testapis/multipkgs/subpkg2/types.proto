syntax = "proto3";

package testapis.multipkgs.subpkg1;

import "testapis/multipkgs/subpkg1/types.proto";

message MessageWithSubpkg {
  testapis.multipkgs.subpkg1.SubpkgMessage message = 1;
  testapis.multipkgs.subpkg1.SubpkgEnum enum = 2;
}
