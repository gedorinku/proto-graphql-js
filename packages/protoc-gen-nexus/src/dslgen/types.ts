import { FieldDescriptorProto } from "google-protobuf/google/protobuf/descriptor_pb";
import { ProtoField, ProtoRegistry } from "../protoTypes";
import { gqlTypeName, isRequiredField } from "./util";

type GqlScalarType = "Int" | "Float" | "String" | "Boolean" | "ID" | "DateTime";

export type GqlItemType =
  | {
      kind: "scalar";
      type: GqlScalarType;
      nullable: boolean;
    }
  | {
      kind: "object";
      type: string;
      nullable: boolean;
    }
  | {
      kind: "enum";
      type: string;
      nullable: boolean;
    };

export type GqlType =
  | GqlItemType
  | {
      kind: "list";
      type: GqlItemType;
      nullable: boolean;
    };

export function detectGqlType(
  f: ProtoField,
  reg: ProtoRegistry,
  opts?: { input?: boolean }
): GqlType {
  if (f.isList()) {
    return {
      kind: "list",
      type: detectGqlItemType(f, reg, opts),
      nullable: false,
    };
  }

  return detectGqlItemType(f, reg, opts);
}

function detectGqlItemType(
  f: ProtoField,
  reg: ProtoRegistry,
  opts?: { input?: boolean }
): GqlItemType {
  const nullable = !isRequiredField(f);
  const pbtype = f.descriptor.getType()!;

  switch (pbtype) {
    case FieldDescriptorProto.Type.TYPE_STRING:
      return { kind: "scalar", type: "String", nullable: false };
    case FieldDescriptorProto.Type.TYPE_DOUBLE:
    case FieldDescriptorProto.Type.TYPE_FLOAT:
      return { kind: "scalar", type: "Float", nullable: false };
    case FieldDescriptorProto.Type.TYPE_INT64:
      return { kind: "scalar", type: "String", nullable: false };
    case FieldDescriptorProto.Type.TYPE_UINT64:
      return { kind: "scalar", type: "String", nullable: false };
    case FieldDescriptorProto.Type.TYPE_INT32:
      return { kind: "scalar", type: "Int", nullable: false };
    case FieldDescriptorProto.Type.TYPE_FIXED64:
      return { kind: "scalar", type: "String", nullable: false };
    case FieldDescriptorProto.Type.TYPE_FIXED32:
      return { kind: "scalar", type: "Int", nullable: false };
    case FieldDescriptorProto.Type.TYPE_UINT32:
      return { kind: "scalar", type: "Int", nullable: false };
    case FieldDescriptorProto.Type.TYPE_SFIXED32:
      return { kind: "scalar", type: "Int", nullable: false };
    case FieldDescriptorProto.Type.TYPE_SFIXED64:
      return { kind: "scalar", type: "String", nullable: false };
    case FieldDescriptorProto.Type.TYPE_SINT32:
      return { kind: "scalar", type: "Int", nullable: false };
    case FieldDescriptorProto.Type.TYPE_SINT64:
      return { kind: "scalar", type: "String", nullable: false };
    case FieldDescriptorProto.Type.TYPE_BOOL:
      return { kind: "scalar", type: "Boolean", nullable: false };
    case FieldDescriptorProto.Type.TYPE_GROUP:
      throw "not supported";
    case FieldDescriptorProto.Type.TYPE_BYTES:
      throw "not supported";
    case FieldDescriptorProto.Type.TYPE_ENUM:
      return {
        kind: "enum",
        type: gqlTypeName(reg.findByFieldDescriptor(f.descriptor)),
        nullable,
      };
    case FieldDescriptorProto.Type.TYPE_MESSAGE:
      switch (f.descriptor.getTypeName()) {
        case ".google.protobuf.Any":
          throw "not supported";
        case ".google.protobuf.BoolValue":
          return {
            kind: "scalar",
            type: "Boolean",
            nullable,
          };
        case ".google.protobuf.BytesValue":
          throw "not supported";
        case ".google.protobuf.DoubleValue":
        case ".google.protobuf.FloatValue":
          return {
            kind: "scalar",
            type: "Float",
            nullable,
          };
        case ".google.protobuf.Duration":
          throw "not implemented";
        case ".google.protobuf.Int32Value":
          return {
            kind: "scalar",
            type: "Int",
            nullable,
          };
        case ".google.protobuf.Int64Value":
          return {
            kind: "scalar",
            type: "String",
            nullable,
          };
        case ".google.protobuf.UInt32Value":
          return {
            kind: "scalar",
            type: "Int",
            nullable,
          };
        case ".google.protobuf.UInt64Value":
          return {
            kind: "scalar",
            type: "String",
            nullable,
          };
        case ".google.protobuf.StringValue":
          return {
            kind: "scalar",
            type: "String",
            nullable,
          };
        case ".google.protobuf.Timestamp":
          return {
            kind: "scalar",
            type: "DateTime",
            nullable,
          };
        default:
          return {
            kind: "object",
            type: gqlTypeName(reg.findByFieldDescriptor(f.descriptor), opts),
            nullable,
          };
      }
    default:
      const _exhaustiveCheck: never = pbtype; // eslint-disable-line
      throw "unreachable";
  }
}
