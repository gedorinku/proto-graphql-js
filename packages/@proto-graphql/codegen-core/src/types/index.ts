export * from "./types";
export { fileLayouts } from "./util";
export type { GenerationParams, LongNumberMapping, PrinterOptions } from "./util";
export * from "./DslFile";
export * from "./ObjectType";
export * from "./ObjectField";
export * from "./ObjectOneofField";
export * from "./InterfaceType";
export * from "./OneofUnionType";
export * from "./SquashedOneofUnionType";
export * from "./ScalarType";
export * from "./EnumType";
export * from "./InputObjectType";
export * from "./InputObjectField";
