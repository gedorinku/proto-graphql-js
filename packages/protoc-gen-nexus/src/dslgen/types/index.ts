export * from "./types";
export type { FullName, GenerationParams } from "./util";
