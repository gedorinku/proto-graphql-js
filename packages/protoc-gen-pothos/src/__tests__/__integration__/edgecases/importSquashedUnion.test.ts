import { testSchemaGeneration } from "../../__helpers__/process.test.helper";

testSchemaGeneration(["edgecases/import_squashed_union/pkg1", "edgecases/import_squashed_union/pkg2"], "ts-proto", {
  schemaTests: [
    [
      "tests query",
      {
        extraSchema: `
          import { builder } from "./builder";
          import { Message as MessageRef } from "./testapis/edgecases/import_squashed_union/pkg2/types_pb_nexus";
          import { Message } from "@testapis/ts-proto/lib/testapis/edgecases/import_squashed_union/pkg2/types";

          builder.queryField("test", (t) => t.field({
            type: MessageRef,
            resolve() {
              return Message.fromPartial({
                msg: {
                  msg1: { body: "hello" },
                },
              });
            },
          }));
      `,
        testQuery: `
          query Test {
            test {
              msg {
                ... on OneofMessage1 {
                  __typename
                  body
                }
              }
            }
          }
      `,
      },
    ],
  ],
});
