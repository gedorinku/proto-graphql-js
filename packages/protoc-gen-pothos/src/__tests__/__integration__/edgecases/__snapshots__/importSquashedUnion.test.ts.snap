// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema generation with ts-proto tests query 1`] = `
Object {
  "data": Object {
    "test": Object {
      "msg": Object {
        "__typename": "OneofMessage1",
        "body": "hello",
      },
    },
  },
}
`;

exports[`schema generation with ts-proto tests query: schema.graphql 1`] = `
"type Message {
  msg: SquashedOneof
}

input MessageInput {
  msg: SquashedOneofInput
}

type OneofMessage1 {
  body: String!
}

input OneofMessage1Input {
  body: String!
}

type Query {
  test: Message!
}

union SquashedOneof = OneofMessage1

input SquashedOneofInput {
  msg1: OneofMessage1Input
}
"
`;
